lizard_auth_client/mixins.py

```diff
@@ -0,0 +1,204 @@
+            self.request.get_full_path(), self.get_login_url(),
+            self.get_redirect_field_name())
+
+
+class RoleRequiredMixin(AccessMixin):
+    """
+    CBV mixin which verifies that the current user has one of the specified
+    roles.
+
+    """
+    role_required = None  # can be a string or a list of strings
+
+    @property
+    def available_roles(self):
+        """
+        Return the available roles.
+
+        If there is a ``SSO_CONNECTED_ROLE_CODE`` or a ``SSO_IGNORE_ROLE_CODES``
+        setting present, exclude those.
+        """
+        # return cached available roles if present
+        if hasattr(self, '_available_roles'):
+            return self._available_roles
+
+        qs = models.Role.objects.all()
+        excluded_roles = []
+        if settings.SSO_CONNECTED_ROLE_CODE:
+            excluded_roles.append(settings.SSO_CONNECTED_ROLE_CODE)
+        if settings.SSO_IGNORE_ROLE_CODES:
+            excluded_roles.extend(settings.SSO_IGNORE_ROLE_CODES)
+        if excluded_roles:
+            qs = qs.exclude(code__in=excluded_roles)
+
+        self._available_roles = qs
+        return self._available_roles
+
+    def get_role_required(self):
+        """
+        Override this method to override the role attribute.
+        Must return an iterable.
+        """
+        if not self.role_required:
+            raise ImproperlyConfigured(
+                '{0} is missing the role attribute. Define {0}.role, or '
+                'override {0}.get_role_required().'.format(
+                    self.__class__.__name__)
+            )
+        if isinstance(self.role_required, six.string_types):
+            roles = (self.role_required,)
+        else:
+            roles = self.role_required
+        return roles
+
+    def has_role(self):
+        """
+        Override this method to customize the way permissions are checked.
+        """
+        user = self.request.user
+        # superusers have all roles implicitly
+        if user.is_superuser:
+            return True
+
+        roles = self.get_role_required()
+        nr_of_user_organisation_roles = models.UserOrganisationRole.objects.\
+            filter(user=user, role__code__in=roles).count()
+        if nr_of_user_organisation_roles > 0:
+            return True
+        else:
+            return False
+
+    @method_decorator(login_required)
+    def dispatch(self, request, *args, **kwargs):
+        if not self.has_role():
+            return self.handle_no_permission()
+        return super(RoleRequiredMixin, self).dispatch(
+            request, *args, **kwargs)
+
+
+class ManagedObjectsMixin(object):
+    """
+    CBV mixin which provides a number of role/permission management-related
+    fields:
+
+    - managed_organisations: the organisations managed by the request user
+    - managed_users: the users managed by the request user for the given
+        organisation(s)
+
+    """
+    def get_managed_organisations(self):
```

Comment:
Dit has_attr/return gebeuren weghalen en vervangen door een `@cached_property`:

    from django.utils.functional import cached_property
    ....

    class ...
 
        @cached_property
        def managed_organisations(self):
            ....
            return xyz

Die doet precies hetzelfde als jouw custom hasattr maar is veel lekkerder leesbaar. En zodra je hem kent ga je hem op 20 andere plekken ook gebruiken :-) Heeeeeeel handig.